### PR TITLE
feat(upgrade): migrate MayastorPool CRs to DiskPool CRs on operator startup

### PR DESCRIPTION
**Pull request by Abhinandan-Purkait**
- #### Restructure the DSP Operator code base
- #### Added the migration from MSP to DSP

### CODE REVIEW

File: k8s/operators/src/diskpool/crd.rs [hunk 1]
```diff
@@ -34,18 +34,21 @@ pub struct DiskPoolSpec {
...
+    /// Create a new DiskPoolSpec from the node and the disks.
+    pub fn new(node: String, disks: Vec<String>) -> Self {
+        Self { node, disks }
+    }
+    /// The node the pool is placed on.
     pub fn node(&self) -> String {
         self.node.clone()
     }
-    /// The disk device the pool is located on
+    /// The disk device the pool is located on.
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/crd.rs]:
```suggestion
    /// The disk devices that compose the pool.
```

File: k8s/operators/src/diskpool/crd.rs [hunk 2]
```diff
@@ -34,18 +34,21 @@ pub struct DiskPoolSpec {
...
     pub fn disks(&self) -> Vec<String> {
         self.disks.clone()
     }
 }
 
 #[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq, JsonSchema)]
-#[non_exhaustive]
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/crd.rs:48]: Why did you remove this?

File: k8s/operators/src/diskpool/main.rs [hunk 3]
```diff
@@ -728,17 +635,39 @@ async fn reconcile(dsp: Arc<DiskPool>, ctx: Arc<OperatorContext>) -> Result<Acti
...
+    match diskpool::ensure_crd(&k8s).await {
+        Ok(o) => {
+            info!(crd = ?o.name_any(), "Created");
+            // let the CRD settle this purely to avoid errors messages in the console
+            // that are harmless but can cause some confusion maybe.
+            tokio::time::sleep(Duration::from_secs(5)).await;
+        }
+
+        Err(e) => {
+            error!("Failed to create CRD error {}", e);
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]:
```suggestion
            error!(%error, "Failed to create CRD");
```

File: k8s/operators/src/diskpool/main.rs [hunk 4]
```diff
@@ -728,17 +635,39 @@ async fn reconcile(dsp: Arc<DiskPool>, ctx: Arc<OperatorContext>) -> Result<Acti
...
+            tokio::time::sleep(Duration::from_secs(1)).await;
+            std::process::exit(1);
+        }
+    }
+
+    // Migrate the MayastorPool CRs to the DiskPool.
+    let _ = migrate_and_clean_msps(&k8s, namespace)
+        .await
+        .map_err(|error| {
+            error!("{error:?}");
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]:
```suggestion
            error!(%error, "Was not able to fully migrate existing v1 CRs");
```

File: k8s/operators/src/diskpool/mod.rs [hunk 5]
```diff
@@ -1,2 +1,11 @@
+//! This module has the definitions and the client specific operations for the
+//! DiskPool CRs. This module is works with the current operator logic.
+
+/// DiskPool client operations.
+mod client;
 /// The DiskPool custom resource definition
-pub mod crd;
+mod crd;
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/mod.rs]: not changed, but can you add the `.` to the previous line

File: k8s/operators/src/lib.rs [hunk 6]
```diff
@@ -1,2 +1,59 @@
...
+
+/// The DiskPool operator.
 pub mod diskpool;
+/// The error module for the operator.
+pub mod error;
+/// The MayaStorPool Operator definitions.
+pub mod mayastorpool;
+
+/// Migrate from MayastorPool
+pub async fn migrate_and_clean_msps(k8s: &Client, namespace: &str) -> Result<(), Error> {
```
Comment (tiagolobocastro) [inline k8s/operators/src/lib.rs]: Shouldn't this be part of the binary rather than the library?

File: k8s/operators/src/lib.rs [hunk 7]
```diff
@@ -1,2 +1,59 @@
...
+    // Check if the MayastorPool CRD is present, and migrate from it if it is.
+    match mayastorpool::check_crd(k8s).await {
+        // Fetch the MayastorPool CRs.
+        Ok(true) => match mayastorpool::list(k8s, namespace, 100).await {
```
Comment (tiagolobocastro) [inline k8s/operators/src/lib.rs]: What's the 100?

File: k8s/operators/src/lib.rs [hunk 8]
```diff
@@ -1,2 +1,59 @@
...
+                        diskpool::DiskPoolSpec::new(node, disks),
+                    )
+                    .await?;
+                }
+                info!("Migration of CRs from MayastorPool to DiskPool complete");
+                // Patch the finalizers and delete the MayastorPool CRs.
+                match mayastorpool::delete(k8s, namespace, &mut msps).await {
+                    Ok(_) => info!("Cleanup of MayastorPool CRs complete"),
+                    Err(error) => {
+                        return Err(Error::Generic {
```
Comment (tiagolobocastro) [inline k8s/operators/src/lib.rs]: Shall we have a non-generic error? Here and below as well?
Also shouldn't we try to migrate and delete all other CR's, otherwise if the first fails we don't try to migrate the other ones?

File: k8s/operators/src/mayastorpool/client.rs [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    loop {
+        let mut result = pools_api.list(&list_params).await?;
+        pools.append(&mut result.items);
+        // Check for the token, if valid then continue.
+        match result.metadata.continue_ {
+            Some(token) if !token.is_empty() => {
+                list_params = list_params.continue_token(token.as_str())
+            }
+            _ => break,
+        };
```
Comment (tiagolobocastro) [inline k8s/operators/src/mayastorpool/client.rs]:
```suggestion
        }
```

File: k8s/operators/src/mayastorpool/client.rs [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    pools_api
+        .delete_collection(&delete_params, &list_params)
+        .await?;
+
+    Ok(())
+}
+
+/// Check whether the MayastorPool crd exists or not.
+pub async fn check_crd(k8s: &Client) -> Result<bool, Error> {
+    let crd_api: Api<k8s_openapi::apiextensions_apiserver::pkg::apis::apiextensions::v1::CustomResourceDefinition> = Api::all(k8s.clone());
```
Comment (tiagolobocastro) [inline k8s/operators/src/mayastorpool/client.rs]: wow that's long, should we shorten it with a use? :D

File: k8s/operators/src/diskpool/crd.rs (hunk 2 above)
Comment (Abhinandan-Purkait) [reply k8s/operators/src/diskpool/crd.rs:48]: Why do we need this? This is causing issues with compilation.

File: k8s/operators/src/lib.rs (hunk 7 above)
Comment (Abhinandan-Purkait) [reply k8s/operators/src/lib.rs]: That's the pagination limit.

File: k8s/operators/src/mayastorpool/client.rs (hunk 10 above)
Comment (Abhinandan-Purkait) [reply k8s/operators/src/mayastorpool/client.rs]: Yeah, we can :-p

File: k8s/operators/src/lib.rs (hunk 8 above)
Comment (Abhinandan-Purkait) [reply k8s/operators/src/lib.rs]: We are deleting after complete migration only. The delete is outside the loop, once migration is done the list of pools we used for migration are passed to delete. If you look into the definition of the delete method. Its a delete_collection call, we aren't deleting one item at a time.

File: k8s/operators/src/lib.rs (hunk 6 above)
Comment (Abhinandan-Purkait) [reply k8s/operators/src/lib.rs]: This is a common code between the two modules. That's why kept in the library. By binary do you mean on main.rs?

File: k8s/operators/src/lib.rs (hunk 7 above)
Comment (Abhinandan-Purkait) [reply k8s/operators/src/lib.rs]: Maybe that should come from some constant. What do you think?

File: k8s/operators/src/mayastorpool/client.rs (hunk 10 above)
Comment (Abhinandan-Purkait) [reply k8s/operators/src/mayastorpool/client.rs]: Done. I did it in the other module but forgot here. 😅

File: k8s/operators/src/lib.rs (hunk 7 above)
Comment (Abhinandan-Purkait) [reply k8s/operators/src/lib.rs]: Added a constant. `PAGINATION_LIMIT`

Comment (tiagolobocastro) [reply k8s/operators/src/lib.rs]: Yeah I saw it was pagination when reviewed the other file :D 
Great, this way it's clearer :+1:

File: k8s/operators/src/diskpool/crd.rs (hunk 2 above)
Comment (tiagolobocastro) [reply k8s/operators/src/diskpool/crd.rs:48]: IIRC this is to ensure that changes to the fields (adding new enum values) are not breaking by ensuring we always use the catch all _ => {}

File: k8s/operators/src/lib.rs (hunk 6 above)
Comment (tiagolobocastro) [reply k8s/operators/src/lib.rs]: So the library is only here for the support tool to inspect the DiskPool CR IIRC. We can move the rest to diskpool which is only used by the diskpool binary IIRC, or perhaps just make sure we don't put the binary modules on the lib?

File: k8s/operators/src/lib.rs (hunk 8 above)
Comment (tiagolobocastro) [reply k8s/operators/src/lib.rs]: Ah ok.. even so, shouldn't we try to delete all of them rather than fail on the first error?

Comment (Abhinandan-Purkait) [reply k8s/operators/src/lib.rs]: As discussed, we keep continuing on errors and the delete would be along with migration.

File: k8s/operators/src/diskpool/crd.rs (hunk 2 above)
Comment (Abhinandan-Purkait) [reply k8s/operators/src/diskpool/crd.rs:48]: As discussed, removing this doesn't break anything.

File: k8s/operators/src/lib.rs (hunk 8 above)
Comment (Abhinandan-Purkait) [reply k8s/operators/src/lib.rs]: Done.

File: k8s/operators/src/lib.rs (hunk 6 above)
Comment (Abhinandan-Purkait) [reply k8s/operators/src/lib.rs]: Done. Moved this to binary. And the only binary module exposed via lib is crd.

File: k8s/operators/src/mayastorpool/client.rs [hunk 11]
```diff
@@ -0,0 +1,82 @@
+use crate::{error::Error, mayastorpool::crd::MayastorPool};
+use k8s_openapi::apiextensions_apiserver::pkg::apis::apiextensions::v1::CustomResourceDefinition;
+use kube::{
+    api::{DeleteParams, ListParams, Patch, PatchParams},
+    Api, Client, CustomResourceExt,
+};
+
+const NOT_FOUND_CODE: u16 = 404;
```
Comment (tiagolobocastro) [inline k8s/operators/src/mayastorpool/client.rs]: Simply use `StatusCode::NOT_FOUND`

File: k8s/operators/src/diskpool/client.rs [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+use kube::{api::PostParams, Api, Client, CustomResourceExt, ResourceExt};
+
+use k8s_operators::{
+    diskpool::crd::{DiskPool, DiskPoolSpec},
+    error::Error,
+    mayastorpool,
+};
+use tracing::{error, info, warn};
+
+const ALREADY_EXISTS_CODE: u16 = 409;
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/client.rs]: Can use `StatusCode::CONFLICT`

File: k8s/operators/src/mayastorpool/client.rs [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+                list_params = list_params.continue_token(token.as_str())
+            }
+            _ => break,
+        }
+    }
+
+    Ok(pools)
+}
+
+/// Delete all the mayastor pool CRs forcefully by patching finalizers.
```
Comment (tiagolobocastro) [inline k8s/operators/src/mayastorpool/client.rs]:
```suggestion
/// Delete the mayastor pool CR forcefully by patching finalizer.
```

File: k8s/operators/src/mayastorpool/crd.rs [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+shortname = "msp",
+printcolumn = r#"{ "name":"node", "type":"string", "description":"node the pool is on", "jsonPath":".spec.node"}"#,
+printcolumn = r#"{ "name":"status", "type":"string", "description":"pool status", "jsonPath":".status.state"}"#,
+printcolumn = r#"{ "name":"capacity", "type":"integer", "format": "int64", "minimum" : "0", "description":"total bytes", "jsonPath":".status.capacity"}"#,
+printcolumn = r#"{ "name":"used", "type":"integer", "format": "int64", "minimum" : "0", "description":"used bytes", "jsonPath":".status.used"}"#,
+printcolumn = r#"{ "name":"available", "type":"integer", "format": "int64", "minimum" : "0", "description":"available bytes", "jsonPath":".status.available"}"#
+)]
+
+/// The pool spec which contains the parameters we use when creating the pool.
+pub struct MayastorPoolSpec {
```
Comment (tiagolobocastro) [inline k8s/operators/src/mayastorpool/crd.rs]: do these need to be pub or is pub(crate) enough? We don't need to expose these to the library?

File: k8s/operators/src/lib.rs [hunk 15]
```diff
@@ -1,2 +1,8 @@
-/// The DiskPool operator
+/// The DiskPool operator.
 pub mod diskpool;
+
+/// Mayastor pool module.
+pub mod mayastorpool;
```
Comment (tiagolobocastro) [inline k8s/operators/src/lib.rs]: We don't need to expose these right?

Comment (Abhinandan-Purkait) [reply k8s/operators/src/lib.rs]: Changed them to pub(crate)

File: k8s/operators/src/mayastorpool/crd.rs (hunk 14 above)
Comment (Abhinandan-Purkait) [reply k8s/operators/src/mayastorpool/crd.rs]: Changed them to pub(crate)

**Review by tiagolobocastro — APPROVED** [review]

**Comment by Abhinandan-Purkait** [comment]
bors merge
